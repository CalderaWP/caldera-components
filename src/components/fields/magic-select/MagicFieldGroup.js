import React from 'react';
import {magicGroupPropTypes} from '../propTypes';
import classNames from 'classnames';
import {MagicItem} from './MagicItem';
import {ButtonGroup} from '../button-group/ButtonGroup';
import {MagicSelect} from './MagicSelect';
import {RenderGroup} from '../../RenderGroup';
import {Message} from '../messages/Message';
import {FieldGroup} from '../FieldGroup';


/**
 * Encapsulates a complete Magic Select field group including label and the type chooser and the input
 */
export class MagicFieldGroup extends React.PureComponent {

	/**
	 * Create a MagicFieldGroup component
	 *
	 * @param {Object} props
	 */
	constructor(props) {
		super(props);
		this.state = {
			currentListType: props.defaultList,
			isOpen: props.isOpen,
		};
		this.onChange = this.onChange.bind(this);
		this.items = this.items.bind(this);
		this.onChange = this.onChange.bind(this);
		this.onSelect = this.onSelect.bind(this);
		this.onInputFocus = this.onInputFocus.bind(this);
		this.renderItem = this.renderItem.bind(this);
		this.onInputBlur = this.onInputBlur.bind(this);
		this.onChangeListType = this.onChangeListType.bind(this);
		this.listTypeOptions = this.listTypeOptions.bind(this);
	}

	/**
	 * Handle direct change events
	 * @param {String|number} newValue
	 */
	onChange(newValue) {
		this.props.onValueChange(newValue);
	}

	/**
	 * Handle when the field gets focus
	 */
	onInputFocus() {
		this.setState({isOpen: true});
	}

	/**
	 * Handle when the field is blurred
	 */
	onInputBlur() {
		this.setState({isOpen: false});
	}


	/**
	 * Handle when the option is chosen
	 * @param {String|number} value
	 */
	onSelect(value) {
		this.props.onValueChange(value);
		this.setState({isOpen: false});
	}

	/**
	 * Update list of tags to show
	 * @param {String}newType
	 */
	onChangeListType(newType){
		if( ! this.state.isOpen ){
			this.setState({isOpen:true});
		}
		this.setState({currentListType:newType});
	}

	/**
	 * Render option
	 *
	 * @param {object} item
	 * @param {bool} isHighlighted
	 * @return {*}
	 */
	renderItem(item, isHighlighted){
		return <MagicItem item={item} isHighlighted={isHighlighted} innerKey={item.innerKey} key={item.innerKey} />;
	}

	/**
	 * Create the list of items
	 *
	 * @return {Array}
	 */
	items() {
		const optionsOrEmpty =(options) => {
			return Array.isArray(options) && options.length ? options : [];
		};

		let items = [];

		if( optionsOrEmpty(this.props.options).length){
			items = optionsOrEmpty(this.props.options);
		}
		else if ('system' === this.state.currentListType ) {
			items = optionsOrEmpty(this.props.systemTagsList);
		} else {
			items = optionsOrEmpty(this.props.fieldsList);
		}

		if (items.length) {
			items.forEach((item, itemIndex) => {
				items[itemIndex].innerKey = `${item.value}-${itemIndex}`;
			});
		}else{
			items.push({
				value: null,
				label: null,
				innerKey:this.props.id
			});
		}

		return items;

	}

	/**
	 * Options for type chooser
	 *
	 * @return {*[]}
	 */
	listTypeOptions(){
		return [
			{
				value: 'fields',
				label: '%',
				ariaLabel: 'Select from field values'
			},
			{
				value: 'system',
				label: '{}',
				ariaLabel: 'Select from system values'
			},
		];
	}



	/**
	 * Render MagicFieldGroup component
	 * @return {*}
	 */
	render() {
		return (
			<div
				className={classNames(
					MagicFieldGroup.classNames.fieldWrapper,
					RenderGroup.classNames.fieldWrapper,
					this.props.className
				)
				}
			>
				<FieldGroup.Label
					id={this.props.id}
					label={this.props.label}
				/>
				{this.props.message.message &&
					<Message
						message={this.props.message}
					/>
				}
				{this.state.isOpen &&
					<ButtonGroup
						onChange={this.onChangeListType}
						options={this.listTypeOptions()}
						value={this.state.currentListType}
					/>
				}

				<MagicSelect
					id={this.props.id}
					options={this.items()}
					onValueChange={this.props.onValueChange}
					value={this.props.value}
					isRequired={this.props.isRequired}
					isOpen={this.state.isOpen}
					onBlur={this.onInputBlur}
					onFocus={this.onInputFocus}
					className={this.props.fieldClassName}
				/>
			</div>
		);
	}
}


/**
 * Prop definitions for MagicFieldGroup component
 */
MagicFieldGroup.propTypes = magicGroupPropTypes;

/**
 * Default property values for MagicFieldGroup component
 *
 * @type {{}}
 */
MagicFieldGroup.defaultProps = {
	defaultList: 'fields',
	isOpen: false,
	message: {
		error: false,
		message: ''
	},
	type: 'magic'
};

/**
 * The names of classes used for HTML elements in MagicFieldGroup component
 * @type {{fieldWrapper: string, input: string, option: string}}
 */
MagicFieldGroup.classNames = {
	fieldWrapper: 'caldera-magic-select-group',
	input: 'caldera-magic-input',
	option: 'caldera-magic-option'
};