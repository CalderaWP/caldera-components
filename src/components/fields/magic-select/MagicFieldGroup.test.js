import renderer from 'react-test-renderer';
import React from 'react';
import {mount, shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {MagicFieldGroup} from './MagicFieldGroup';

Enzyme.configure({adapter: new Adapter()});

const genericChangeHandler = () => {
};
describe('MagicFieldGroup component', () => {


	it( 'matches snapshot', () => {
		const component = renderer.create(
			<MagicFieldGroup
				id={'magic-3'}
				label={'Hi Roy'}
				fieldClassName={'magic'}
				onValueChange={genericChangeHandler}
				options={[
					{
						label: 'HTML',
						value: 'html'
					},
					{
						label: 'Plain Text',
						value: 'plain'
					}
				]}
				isOpen={false}
			/>
		);
		expect(component.toJSON() ).toMatchSnapshot();
	});

	describe( 'Inner input', () => {
		it('Has inner input', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-1'}
					fieldClassName={'magic'}
					label={'Hi Roy'}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: 'HTML',
							value: 'html'
						},
						{
							label: 'Plain Text',
							value: 'plain'
						}
					]}
					isOpen={false}
				/>
			);
			expect(component.find('input').length).toBe(1);
		});
	});
	describe('Options', () => {
		it('shows no options if closed', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-3'}
					label={'Hi Roy'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: 'HTML',
							value: 'html'
						},
						{
							label: 'Plain Text',
							value: 'plain'
						}
					]}
					isOpen={false}
				/>
			);
			expect(component.find('.magic-input-option')).toHaveLength(0);
		});


		it('Uses options prop by default - right number of options', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-4'}
					label={'Hi Roy'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: 'HTML',
							value: 'html'
						},
						{
							label: 'Plain Text',
							value: 'plain'
						}
					]}
					isOpen={true}
				/>
			);
			component.find('input').simulate('focus');
			expect(component.find('.magic-input-option')).toHaveLength(2);
		});

		it('Uses options prop by default', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-4'}
					label={'Hi Roy'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: 'HTML',
							value: 'html'
						},
						{
							label: 'Plain Text',
							value: 'plain'
						}
					]}
					isOpen={true}
				/>
			);
			component.find('input').simulate('focus');
			expect(component.find('.magic-input-option').length).toBe(2);
		});

		it('Uses fieldsList prop if no options prop', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-5'}
					label={'Hi Roy'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					fieldsList={[
						{
							label: '0',
							value: 0
						},
						{
							label: '1',
							value: 1
						},
						{
							label: '3',
							value: 3
						}
					]}
					systemTagsList={[
						{
							label: '3',
							value: 3
						}
					]}
					isOpen={true}
				/>
			);
			component.find('input').simulate('focus');
			expect(component.find('.magic-input-option').length).toBe(3);
		});

		it('Uses systemTagsList prop if no options prop and currentListType state is system', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-5'}
					fieldClassName={'magic'}
					label={'Hi Roy'}
					onValueChange={genericChangeHandler}
					fieldsList={[
						{
							label: '0',
							value: 0
						},
						{
							label: '1',
							value: 1
						},
						{
							label: '3',
							value: 3
						}
					]}
					systemTagsList={[
						{
							label: '3',
							value: 3
						}
					]}
					isOpen={true}
				/>
			);
			component.find('input').simulate('focus');
			component.setState({currentListType: 'system'});
			expect(component.find('.magic-input-option').length).toBe(1);
		});

		it('Receives updated value', () => {
			let value = '1';
			const component = mount(
				<MagicFieldGroup
					id={'magic-5'}
					fieldClassName={'magic'}
					label={'Hi Roy'}
					onValueChange={(newValue) => {
						value = newValue;
					}}
					fieldsList={[
						{
							label: '0',
							value: 0
						},
						{
							label: '1',
							value: 1
						},
						{
							label: '3',
							value: 3
						}
					]}
					systemTagsList={[
						{
							label: '3',
							value: 3
						}
					]}
					value={value}
					isOpen={true}
				/>
			);
			component.find('input').simulate('focus');
			component.find('input').simulate('change', { target: { value: '3' } });
			expect( value ).toEqual('3');
		});
	});


	describe('Value', () => {
		it('Uses props.value to set value', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-7'}
					fieldClassName={'magic'}
					label={'Hi Roy'}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: '1',
							value: 1
						}
					]}
					value={1}
				/>
			);
			expect(component.find('input').prop('value')).toBe(1);
		});


		it('Passes updated value properly through the onSelect handler', () => {
			let updatedValue = null;
			const component = mount(
				<MagicFieldGroup
					id={'magic-9'}
					fieldClassName={'magic'}
					label={'Hi Roy'}
					onValueChange={(newValue) => {
						updatedValue = newValue;
					}}
					options={[
						{
							label: '1',
							value: 1
						},
						{
							label: '14',
							value: 14
						}
					]}
					value={1}
				/>
			);

			component.instance().onSelect(14);
			expect(updatedValue).toEqual(14);
		});
	});

	describe( 'Button group for type choice', () => {
		it( 'Outputs the buttons if open', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-50'}
					fieldClassName={'magic'}
					label={'Hi Roy'}
					onValueChange={genericChangeHandler}
					fieldsList={[
						{
							label: '0',
							value: 0
						},
						{
							label: '1',
							value: 1
						},
						{
							label: '3',
							value: 3
						}
					]}
					systemTagsList={[
						{
							label: '3',
							value: 3
						}
					]}
					isOpen={true}
				/>
			);
			expect(component.find('button').length).toBe(2);
		});
		it( 'Does not output the buttons if not open', () => {
			const component = mount(
				<MagicFieldGroup
					id={'magic-50'}
					fieldClassName={'magic'}
					label={'Hi Roy'}
					onValueChange={genericChangeHandler}
					fieldsList={[
						{
							label: '0',
							value: 0
						},
						{
							label: '1',
							value: 1
						},
						{
							label: '3',
							value: 3
						}
					]}
					systemTagsList={[
						{
							label: '3',
							value: 3
						}
					]}
					isOpen={false}
				/>
			);
			expect(component.find('button').length).toBe(0);
		});
	});
});