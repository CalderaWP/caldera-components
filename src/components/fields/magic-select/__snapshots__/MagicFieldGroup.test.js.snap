// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicFieldGroup component matches snapshot 1`] = `
<div
  className="caldera-magic-select-group caldera-config-field"
>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <input
      aria-autocomplete="list"
      aria-expanded={false}
      autoComplete="off"
      className="caldera-magic-input caldera-magic-input-magic-3 magic"
      id="magic-3"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="combobox"
      value=""
    />
  </div>
</div>
`;
